### PR TITLE
libvirt: improve the get domain IPs mechanism

### PR DESCRIPTION
**Pull request by wainersm**
Currently after the Libvirt domain is created it will wait for a fixed amount of time (30s) before query libvirt for the domain's IP addresses. This is clearly sub-optimal.

Moreover, while working on the CI job for libvirt (https://github.com/confidential-containers/cloud-api-adaptor/pull/1318) I realized 30s is not enough if the job runner is in an Azure VM, but if I increase the timeout too much....even worse.

So I am sending those changes the refactor the code to get the IP addresses. It will now be polling the domain every 3 seconds and retry up to 20 times. So approximately 60s of wait time on worst case.

### CODE REVIEW

**Comment by wainersm** [comment]
@snir911 could you please review this? :)

**Comment by wainersm** [comment]
Tests are passing:

```
=== RUN   TestLibvirtCreateSimplePod
=== RUN   TestLibvirtCreateSimplePod/SimplePeerPod_test
    common_suite_test.go:160: Expected Pod State: Running
    common_suite_test.go:161: Current Pod State: Running
=== RUN   TestLibvirtCreateSimplePod/SimplePeerPod_test/PodVM_is_created
time="2023-08-17T16:19:54-03:00" level=info msg="Deleting pod... nginx"
--- PASS: TestLibvirtCreateSimplePod (140.14s)
    --- PASS: TestLibvirtCreateSimplePod/SimplePeerPod_test (140.14s)
        --- PASS: TestLibvirtCreateSimplePod/SimplePeerPod_test/PodVM_is_created (0.07s)
=== RUN   TestLibvirtCreatePodWithConfigMap
=== RUN   TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test
    common_suite_test.go:160: Expected Pod State: Running
    common_suite_test.go:161: Current Pod State: Running
=== RUN   TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test/Configmap_is_created_and_contains_data
time="2023-08-17T16:21:24-03:00" level=info msg="Data Inside Configmap: Hello, world"
time="2023-08-17T16:21:24-03:00" level=info msg="Deleting Configmap... nginx-config"
time="2023-08-17T16:21:24-03:00" level=info msg="Deleting pod... nginx-configmap-pod"
--- PASS: TestLibvirtCreatePodWithConfigMap (90.26s)
    --- PASS: TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test (90.26s)
        --- PASS: TestLibvirtCreatePodWithConfigMap/ConfigMapPeerPod_test/Configmap_is_created_and_contains_data (5.16s)
=== RUN   TestLibvirtCreatePodWithSecret
=== RUN   TestLibvirtCreatePodWithSecret/SecretPeerPod_test
    common_suite_test.go:160: Expected Pod State: Running
    common_suite_test.go:161: Current Pod State: Running
=== RUN   TestLibvirtCreatePodWithSecret/SecretPeerPod_test/Secret_has_been_created_and_contains_data
time="2023-08-17T16:22:29-03:00" level=info msg="Username from secret inside pod: admin"
time="2023-08-17T16:22:35-03:00" level=info msg="Password from secret inside pod: password"
time="2023-08-17T16:22:35-03:00" level=info msg="Deleting Secret... nginx-secret"
time="2023-08-17T16:22:35-03:00" level=info msg="Deleting pod... nginx-secret-pod"
--- PASS: TestLibvirtCreatePodWithSecret (70.25s)
    --- PASS: TestLibvirtCreatePodWithSecret/SecretPeerPod_test (70.25s)
        --- PASS: TestLibvirtCreatePodWithSecret/SecretPeerPod_test/Secret_has_been_created_and_contains_data (10.19s)
PASS
ok      github.com/confidential-containers/cloud-api-adaptor/test/e2e   351.159s
```

**Comment by portersrc** [comment]
Hey @wainersm LGTM!

**Comment by snir911** [comment]
@wainersm Why would podvm will have more than one ip? and is it possible we'll get the wrong IP in case it has it
overall LGTM i just didn't get this point, thanks!

File: pkg/adaptor/cloud/libvirt/libvirt.go [hunk 1]
```diff
@@ -474,25 +505,23 @@ func CreateDomain(ctx context.Context, libvirtClient *libvirtClient, v *vmConfig
 	logger.Printf("VM id %s", v.instanceId)
 
 	// Wait for sometime for the IP to be visible
-	// TBD: Figure out a better mechanism
-	time.Sleep(30 * time.Second)
+	for retry := 1; retry <= GetDomainIPsRetries; retry++ {
```
[the remark this answers is not in the document]
Comment (wainersm) [reply pkg/adaptor/cloud/libvirt/libvirt.go:508]: Interesting, never used that lib. Let me give it a try.

**Comment by wainersm** [comment]
hi @snir911 !

> @wainersm Why would podvm will have more than one ip? and is it possible we'll get the wrong IP in case it has it overall LGTM i just didn't get this point, thanks!

I don't think the podvm will have more than one IP but the `getDomainIPs()` function will return all IPs just in case. I expect that the caller, in this case the code wiring the network, is going to deal with multiple IPs if the case.

**Review by snir911 — APPROVED** [review]

**Review by bpradipt — APPROVED** [review]
/lgtm
Thanks @wainersm

**Pull request merged**